Move function calls with static arguments out of loops

mitchute opened the pull request:
See issue #5832 for an example. Review the code and cleanup these issues.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this

Commit:
Make some constants 'const's

Commit:
Pull constants out of loops and move to vector which are calculated once upon init

Commit:
Fix iterator range for new const vectors

rraustad commented:
36% improvement. Outstanding.

Myoldmopar commented:
> 36% improvement. Outstanding.

Please hold, a bunch of those performance files failed, so they are probably showing up as zeroes.  I'm definitely hopeful for good improvement from @mitchute but that might be a bit too high.

JasonGlazer commented:
@mitchute thanks for taking this one on. I will be happy to review this. Hopefully most of these CI issue will be easy to fix.

mitchute commented:
I'm hopeful this next round of CI results will give some insight into the savings.

Commit:
More cleanup

rraustad commented:
This is the type of err change that says something is very different.
I forgot the specific file name now but I think it was DOAToFanCoilInlet

        ************* SizeHVACSingleDuct: Potential issue with equipment sizing for AirTerminal:SingleDuct:VAV:NoReheat = "SPACE1-1 DOAS AIR TERMINAL".
        **   ~~~   ** User-Specified Minimum Cooling Air Flow Fraction of 1.00000 [m3/s]
    -   **   ~~~   ** differs from Design Size Minimum Cooling Air Flow Fraction of 0.72766 [m3/s]
    +   **   ~~~   ** differs from Design Size Minimum Cooling Air Flow Fraction of 0.00000 [m3/s]

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into RemoveSinCosSolarShading

rraustad commented:
Pardon me for interrupting. I merged develop to see if that eased the diff's.

mitchute commented:
Thanks for merging that in @rraustad.

Commit:
Another pass looking for std::cos. Changes in GroundHeatExchangers.cc, SolarReflectionManager.cc, and WindowManager.cc

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

Commit:
Standardize changes

Commit:
Work in DaylightingDevices.cc

Commit:
Space padding in DaylightingManager.cc

rraustad commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -3691,12 +3691,12 @@ namespace SolarShading {
-			for ( int IPhi = 1; IPhi <= NPhi; ++IPhi ) { // Loop over patch altitude values
-				SUNCOS( 3 ) = SinPhi[ IPhi ];
+			for ( int IPhi = 0; IPhi < NPhi; ++IPhi ) { // Loop over patch altitude values
+				SUNCOS( 3 ) = SinPhi[IPhi];
```
any requirement for spacing here? e.g., SinPhi[ IPhi ]

mitchute replied:
Probably should add them, but I'd guess that @Myoldmopar has the final say. As with everything, there's a variety with and without throughout the code.

Myoldmopar replied:
I prefer appropriate padding.  The convention EnergyPlus adopted from the translation is to pad everything.  Perhaps we can make a huge push to clear out PRs after the release and make a big move to finalize our style and get a style checker to verify pushed code passes.

rraustad commented on src/EnergyPlus/TarcogShading.cc:
```diff
@@ -588,6 +588,7 @@ namespace TarcogShading {
 		Real64 const s1_2( pow_2( s1 ) );
 		Real64 const s2_2( pow_2( s2 ) );
 		Real64 const s1_s2_2( pow_2( s1 / s2 ) );
+		Real64 const CosTilt = std::cos( tilt );
```
I'm no expert but I see a difference in initialization syntax, e.g., x( y ) versus x = y.

mitchute replied:
I've been trying to keep it consistent by going with ```Real64 cos_foo( std::cos( foo ) );```. I just missed this one. Again, any preference, @Myoldmopar ?

Myoldmopar replied:
This is another one that I differ from the adopted convention.  I prefer the equals, and there are subtle differences, but not for these specific calls.  

These types of things alone don't always hold up a PR from going in, as our style guide doesn't even match the code, which results in some spots of the code adopting new patterns.

Commit:
Formatting. Another pass through looking for std::sin. Modified GroundHeatExchangers.cc, HeatBalanceHAMTManager.cc, SolarCollectors.cc, SolarReflectionManager.cc, TarcogShading.cc, and WindowManager.cc.

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

Commit:
Fix variable initialization formatting. New changes in EcoRoofManager.cc and FuelCellElectricGenerator.cc

Commit:
Fix warning message

rraustad commented:
My first thought is that there should be no diff's. I have, at times, thought that a compilation of different code may show minor diff's (E-15), but major diff's should not be present here. Do you agree?

JasonGlazer commented on src/EnergyPlus/AirflowNetworkSolver.cc:
```diff
@@ -1469,9 +1469,9 @@ namespace AirflowNetworkSolver {
 					S2 = std::sqrt( 2.0 * RHOZ( n ) * PDROP ) * DisSysCompDuctData( CompNum ).A;
 					FTT = S2 / std::sqrt( ld / pow_2( g ) + DisSysCompDuctData( CompNum ).TurDynCoef );
 					if ( LIST >= 4 ) gio::write( Unit21, Format_901 ) << " dwt:" << i << S2 << FTT << g;
+					B = ( 9.3 * VISCZ( n ) * DisSysCompDuctData( CompNum ).A ) / ( FT * DisSysCompDuctData( CompNum ).Rough );
```
This line should not be moved out of the while loop. It depends on FT which depends on FTT which changes each time through the loop.

JasonGlazer replied:
Thanks for fixing.

JasonGlazer commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -606,21 +606,26 @@ namespace EcoRoofManager {
-				P2 = 4.0 * ( sigmaf * epsilonf * epsilong * Sigma ) * pow_3( SoilTK ) / EpsilonOne + 0.1 * sigmaf * sheatf + LAI * Rhoaf * Cf * Lef * Waf * rn * ( 0.1 * sigmaf * Mg ) / dOne * dqg;
-				P3 = 4.0 * ( -sigmaf * epsilonf * Sigma - ( sigmaf * epsilonf * Sigma * epsilong ) / EpsilonOne ) * pow_3( LeafTK ) + ( 0.6 * sigmaf - 1.0 ) * sheatf + LAI * Rhoaf * Cf * Lef * Waf * rn * ( ( ( 0.6 * sigmaf * rn ) / dOne ) - 1.0 ) * dqf;
+				P1 = sigmaf * ( RS * ( 1.0 - Alphaf ) + epsilonf * Latm ) - 3.0 * sigmaf * epsilonf * epsilong * Sigma * pow_4_SoilTK / EpsilonOne - 3.0 * ( -sigmaf * epsilonf * Sigma - sigmaf * epsilonf * epsilong * Sigma / EpsilonOne ) * pow_4_LeafTK + sheatf * ( 1.0 - 0.7 * sigmaf ) * ( Ta + KelvinConv ) + LAI * Rhoaf * Cf * Lef * Waf * rn * ( ( 1.0 - 0.7 * sigmaf ) / dOne ) * qa + LAI * Rhoaf * Cf * Lef  ...
+				P2 = 4.0 * ( sigmaf * epsilonf * epsilong * Sigma ) * pow_3_SoilTK / EpsilonOne + 0.1 * sigmaf * sheatf + LAI * Rhoaf * Cf * Lef * Waf * rn * ( 0.1 * sigmaf * Mg ) / dOne * dqg;
```
I don't think pow_3_SoilTK or pow_4_SoilTK can be precalculated outside of the loop because later on it is computed again within the loop

JasonGlazer replied:
Thanks for fixing

JasonGlazer commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -606,21 +606,26 @@ namespace EcoRoofManager {
+				T2G = 4.0 * ( -( 1.0 - sigmaf ) * epsilong * Sigma - sigmaf * epsilonf * epsilong * Sigma / EpsilonOne ) * pow_3_SoilTK + ( 0.1 * sigmaf - 1.0 ) * sheatg + Rhoag * Ce * Leg * Waf * Mg * ( 0.1 * sigmaf * Mg / dOne - Mg ) * dqg - Qsoilpart2;
 
-				T3G = ( 4.0 * ( sigmaf * epsilong * epsilonf * Sigma ) / EpsilonOne ) * pow_3( LeafTK ) + 0.6 * sigmaf * sheatg + Rhoag * Ce * Leg * Waf * Mg * ( 0.6 * sigmaf * rn / dOne ) * dqf;
+				T3G = ( 4.0 * ( sigmaf * epsilong * epsilonf * Sigma ) / EpsilonOne ) * pow_3_LeafTK + 0.6 * sigmaf * sheatg + Rhoag * Ce * Leg * Waf * Mg * ( 0.6 * sigmaf * rn / dOne ) * dqf;
```
I don't think pow_3_LeafTK or pow_4_LeafTK can be precalculated outside of the loop because later on LeafTK is computed again within the loop

JasonGlazer replied:
Thanks for fixing

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2005,6 +2005,13 @@ namespace FuelCellElectricGenerator {
 		Tkel = ( FluidTemp + KelvinConv );
 		Tsho = ( FluidTemp + KelvinConv ) / 1000.0;
```
Tkel and Tsho should be const

JasonGlazer replied:
OK

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2103,6 +2110,13 @@ namespace FuelCellElectricGenerator {
 
 		tempHair = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Tsho and Tkel should be const

JasonGlazer replied:
ok

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2200,6 +2214,12 @@ namespace FuelCellElectricGenerator {
 
 		tempCp = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Tsho and Tkel should be const

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2297,6 +2317,13 @@ namespace FuelCellElectricGenerator {
 
 		tempHfuel = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Same as above.

JasonGlazer replied:
ok

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2397,6 +2424,13 @@ namespace FuelCellElectricGenerator {
 
 		tempHprodGases = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Same as above.

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2502,6 +2536,13 @@ namespace FuelCellElectricGenerator {
 
 		tempCp = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Same.

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2609,6 +2650,13 @@ namespace FuelCellElectricGenerator {
 
 		tempCp = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Same.

JasonGlazer commented on src/EnergyPlus/FuelCellElectricGenerator.cc:
```diff
@@ -2716,6 +2764,13 @@ namespace FuelCellElectricGenerator {
 
 		tempCp = 0.0;
 
+		Real64 const pow_2_Tsho( pow_2( Tsho ) );
```
Same

JasonGlazer commented on src/EnergyPlus/GroundHeatExchangers.cc:
```diff
@@ -514,28 +514,21 @@ namespace GroundHeatExchangers {
+		Real64 pipeOuterRadius = pipeOutDia / 2.0;
 
-		pipeOuterRadius = pipeOutDia / 2.0;
+		Real64 const cos_theta = std::cos( theta );
```
These changes are ok but since they are not in a loop they will probably not have much performance impact.

JasonGlazer commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -6277,21 +6284,19 @@ Label20: ;
-					tvisFit( IPhi ) = POLYF( CosPhi, Construct( ConstrNum ).TransVisBeamCoef );
-					rfsolFit( IPhi ) = POLYF( CosPhi, Construct( ConstrNum ).ReflSolBeamFrontCoef );
+				for ( IPhi = 0; IPhi < 10; ++IPhi ) {
+					if ( std::abs( CosPhi[IPhi] ) < 0.0001 ) CosPhi[IPhi] = 0.0;
```
This line could also be moved to be precalculated since CosPhi[IPhi} is a set of constants

JasonGlazer commented on src/EnergyPlus/SolarReflectionManager.cc:
```diff
@@ -1145,27 +1149,37 @@ namespace SolarReflectionManager {
+			sin_Theta.push_back( std::sin( Theta ) );
+			cos_Theta.push_back( std::cos( Theta ) );
+		}
+
```
Great catch!

JasonGlazer commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -522,6 +535,19 @@ namespace SolarShading {
 
 		}
 
+		// Initialize these once
```
Great idea since they are used all over this file.

JasonGlazer commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -8450,22 +8442,14 @@ namespace SolarShading {
-			CosPhi = std::cos( Phi );
+			for ( int ITheta = 0; ITheta < NTheta; ++ITheta ) { // Loop over patch azimuth values
+				SUNCOS( 1 ) = cos_Phi[IPhi] * cos_Theta[ITheta];
+				SUNCOS( 2 ) = cos_Phi[IPhi] * sin_Theta[ITheta];
```
Thanks!

JasonGlazer commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -7945,13 +7943,16 @@ namespace WindowManager {
-							//         Integrate transmittance using coordiante transform
-							RelativeAzimuth = std::asin( std::sin( SunAltitude ) * std::cos( SunAzimuth ) ); // phi prime
-							RelativeAltitude = std::atan( std::tan( SunAltitude ) * std::sin( SunAzimuth ) ); // alpha
+							Real64 SunAltitude = ( 90.0 / M ) * ( i - 1 ) * ( Pi / 180.0 ); // Altitude angle of sun during integration
```
SunAzimuth, SunAltitude, and SkyArea could all be precalculated outside of the SurfNum loop since N and M are constants. So could RelativeAzimuth and RelativeAltitude

JasonGlazer replied:
Looks good.

JasonGlazer commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -8007,8 +8008,8 @@ namespace WindowManager {
 						for ( i = 90 / Material( MatNum ).ScreenMapResolution + 1; i >= 1; --i ) {
-							SunAzimuth = Material( MatNum ).ScreenMapResolution * ( j - 1 ) * ( Pi / 180.0 );
-							SunAltitude = Material( MatNum ).ScreenMapResolution * ( i - 1 ) * ( Pi / 180.0 );
+							Real64 SunAzimuth = Material( MatNum ).ScreenMapResolution * ( j - 1 ) * ( Pi / 180.0 );
```
Should also make Pi/180 a constant. I think we might have one for that already.

JasonGlazer commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -8231,7 +8232,9 @@ Label99999: ;
-					if ( ( phib > 0.0 && phib <= PiOvr2 && phis <= phib ) || ( phib > PiOvr2 && phib <= Pi && phis > -( Pi - phib ) ) ) fEdge1 = Blind( BlindNum ).SlatThickness * std::abs( std::sin( gamma ) ) / ( ( Blind( BlindNum ).SlatSeparation + Blind( BlindNum ).SlatThickness / std::abs( std::sin( phib ) ) ) * std::cos( phis ) );
+					if ( ( phib > 0.0 && phib <= PiOvr2 && phis <= phib ) || ( phib > PiOvr2 && phib <= Pi && phis > -( Pi - phib ) ) ) {
+						fEdge1 = Blind( BlindNum ).SlatThickness * std::abs( std::sin( gamma ) ) / ( ( Blind( BlindNum ).SlatSeparation + Blind( BlindNum ).SlatThickness / std::abs( std::sin( phib ) ) ) * std::cos( phis ) );
+					}
```
Thank you. I hate if's like that.

JasonGlazer reviewed (changes requested):
Overall I appreciate you going so much farther than just fixing the SkyDifSolarShading() issue with sin and cos. I did find a couple of issues that need to be fixed.

Commit:
Merge remote-tracking branch 'origin/RemoveSinCosSolarShading' into RemoveSinCosSolarShading

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

Commit:
Address diffs and change requests by @JasonGlazer and @rraustad

mitchute commented:
@JasonGlazer @rraustad Thanks for your input. I think this last commit addresses your comments. Integration tests ran successfully on my local machine, but let's see what CI comes back with.

Commit:
Fix WindowManager

Commit:
Fix daylighting and windows diffs. Add EP_UNUSED to unused variables.

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

Commit:
Fix warnings about unused variables

Commit:
Merge remote-tracking branch 'origin/develop' into RemoveSinCosSolarShading

Commit:
Fix merge errors

JasonGlazer commented:
@mitchute let me know when the code changes are over and I will finish up the review.

Commit:
remove spurious files created by gitKraken after merge conflict

Commit:
One more change to DaylightingDevices.cc to cleanup diffs

Commit:
Fix array range errors in HeatBalanceManager.cc

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

mitchute commented:
@JasonGlazer I think this is going to be it. We'll see what CI comes back with, but I'm not expecting any more changes.

JasonGlazer reviewed (approved):
Looks good!

Commit:
Merge branch 'develop' into RemoveSinCosSolarShading

mitchute commented:
@Myoldmopar I just merge develop in one more time to see if these tests get cleaned up. I'm not seeing any failures on my machines (Windows 7 and Mac), but let me know if you see anything else on this.

JasonGlazer commented:
@mitchute it still seems to be having some small problems with the CI in a few places. Are you planning on looking into those more?

mitchute commented:
@JasonGlazer All of those run fine locally on my machine, so I'm not sure what else to do at this point.

mjwitte commented:
@JasonGlazer The failure you saw here for 5ZoneAirCooledWithDOASAirLoop on the linux integration coverage tests is also failing in develop.  So it's not an issue with this branch.

JasonGlazer committed:
Merge branch 'develop' into RemoveSinCosSolarShading
